C++/source/number_theory/fibonacci_efficient.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+    Fibonacci number
+    -----------------
+    Find the N-th Fibonacci number, given the value of N.
+    (The Fibonacci sequence is characterized by the fact that every number
+    after the first two is the sum of the two preceding ones.)
+
+    Time complexity
+    ---------------
+    O(log N), where N is the term of the Fibonacci sequence to calculate
+
+    Space complexity
+    ----------------
+    O(1).
+*/
+
+#include <iostream>
+#include <vector>
+
+using namespace std;
+
+typedef unsigned long long int ULL;
+
+const int MAX_N = 93;   // fibonacci(94) goes beyond the range of ULL
+
+// Binary exponentiation (of matrix)
+vector< vector<ULL> > pow_matrix(const vector< vector<ULL> > &matrix, const int n) {
+    if (n == 1)
+        return matrix;
+    else if (n % 2 == 0) {
+        // a ^ n = a ^ (n / 2) * a ^ (n / 2)
+        vector< vector<ULL> > tmp_matrix = pow_matrix(matrix, n / 2);
+        vector< vector<ULL> > sqr_tmp_matrix(2, vector<ULL> (2, 0));
+
+        // matrix product
+        for (int i = 0; i < 2; i++)
+            for (int j = 0; j < 2; j++)
+                for (int k = 0; k < 2; k++)
+                    sqr_tmp_matrix[i][j] += tmp_matrix[i][k] * tmp_matrix[k][j];
+        return sqr_tmp_matrix;
+    }
+    else {
+        // a ^ n = a * a ^ (n - 1)
+        vector< vector<ULL> > tmp_matrix = pow_matrix(matrix, n - 1);
+        vector< vector<ULL> > sqr_tmp_matrix(2, vector<ULL> (2, 0));
+
+        // matrix product
+        for (int i = 0; i < 2; i++)
```

Comment:
Since these lines repeat themselves twice, the function can be simplified to:
```C++
if (n == 1)
    return matrix;

vector<vector<ULL>> tmp_matrix;
if (n % 2 == 0)
    tmp_matrix = pow_matrix(matrix, n / 2);
else
    tmp_matrix = pow_matrix(matrix, n - 1);

array<array<2, ULL>> sqr_tmp_matrix;
for (size_t i = 0; i < 2; i++)
    for (size_t j = 0; j < 2; j++)
        for (size_t k = 0; k < 2; k++)
            sqr_tmp_matrix[i][j] += tmp_matrix[i][k] * tmp_matrix[k][j];

return sqr_tmp_matrix;
```

Reply:
There is a different multiplication of matrices:
sqr_tmp_matrix[i][j] += tmp_matrix[i][k] * tmp_matrix[k][j]; in n / 2 part
and
sqr_tmp_matrix[i][j] += tmp_matrix[i][k] * matrix[k][j]; in n - 1 part

Reply:
I can write a separate function for that.